README.md
zero factory
===========

This is the zeroMQ app factory for Pressyo products. Multiple language is supported. Below are deployment instructions

# Python #

## Requirements ##
gevent is now a required package. gevent 1.0rc1 is recommended. 

Due to the fact that gevent 1.0 is yet to be in the PyPI, you have to manually install it first:

1. Install libev4: ` sudo apt-get install libev4 libev-libevent-dev `
2. Download the latest version of gevent (gevent-1.0rc1): ` wget https://github.com/downloads/SiteSupport/gevent/gevent-1.0rc1.tar.gz `
3. Untar it: `tar -xzvf gevent-1.0rc1.tar.gz`
4. Go to the untar'd directory, then make: `python setup.py build`
5. Install it (sudo may be required): `python setup.py install`

## Installing zero-factory ##
Good news everybody, you can actually install using pip! This is how you'd do it:

## Deploying without a build server ##
1. clone the git repo into local system
2. To install: `sudo pip install -e /path/to/zeroFactory/python/ZeroFactory-x.y.z.tar.gz`
3. To use: `from zeroFactory import appFactory`
4. Usage instructions can be found in the docs

## Deploying with a build server (or your own cheeseshop) ##
1. clone the git repo into the build server
2. Navigate to `/path/to/zeroFactory/python`
3. Run `python setup.py sdist`
4. Register your app with your own cheeseshop: see [Tarek Ziade's excellent guide](http://ziade.org/2008/03/20/how-to-run-your-own-private-pypi-cheeseshop-server/). You're on your own if you use Chishop instead
5. To install on slave servers: `pip install zeroFactory -i http://cheeseshop.workfloo.com`
6. Usage instructions can be found in the docs.


# Javascript (node.js) #
... stuff here ..



python/CHANGES.txt
v0.1.1, 20/10/2012 -- Initial release.
v0.1.2, 20/10/2012 -- Updated to have proper error messages
v0.1.3, 20/10/2012 -- Bugfix  on errorDict
v0.1.4, 20/10/2012 -- Bugfix  on errorDict - error messages can now be properly serialized
v0.1.5, 20/10/2012 -- uses proper serializer API for msgpack. Reduces try/except blocks
v0.1.6, 21/10/2012 -- Major bugfixes: success and error messages are now correct, and there are tests for those now
v0.1.7, 21/10/2012 -- Updated to support pyzmq 2.2.0.1 which has native greenlet and gevent support. Not sure why yet
v0.1.8, 23/10/2012 -- Changed exceptions.py into zeroFactoryExceptions.py as it could interfere with the native python module `exceptions`
v0.1.9, 24/10/2012 -- Wrapped up successful results with SUCCESSMESSAGE. Now methods merely have to return a dict of their results.
v0.1.10, 28/10/2012 -- Bug fix on Line 71: UnboundLocalError. Now it should have correct error strings
v0.1.11, 28/10/2012 -- Added self.verbose. To aid people with debugging. Also temporary logging : PRINT ALL THE THINGS!
v0.1.12, 07/11/2012 -- Used zmq.green instead of zmq. This changes the requirement to require gevent. Recommended: geven1.0rc1
v0.1.13, 09/11/2012 -- Added appWrapper. I figure if you provide a factory, might as well provide a standardized client.
v0.1.14, 09/11/2012 -- Added JSONRPCMESSAGE as a constant in appWrapper for easier and shared access.
v0.1.15, 10/11/2012 -- Added a method: self.bind(). This makes bad sockets disconnects much easier
v0.1.16, 11/11/2012 -- Added self.run() to reconnect info as it would have just reconnect, and then not run
v0.1.17, 13/11/2012 -- Changed the verbose error code. Now it will output traceback if verbose.
v0.1.18, 26/11/2012 -- Added better error handling. Most general errors will now throw error code -32603 and the app will not randomly restart
v0.1.19, 27/11/2012 -- Added error handling for when no result was found. It will now reply and raise -32404, which is 404 not found
v0.1.20, 30/11/2012 -- Better error handling. Replaced the old except zmq.core.error.ZMQError handler since that exception no longer exists.
v0.1.21, 13/01/2013 -- Made it that only 'rep' type sockets are replying. Else it'd stay silent

python/test/appTest.py
import sys
import os

# fix import and weird WSGI path issues
abspath = os.path.abspath(os.path.dirname(sys.argv[0]))
parentdir = os.path.dirname(abspath)
sys.path.insert(1, parentdir)  # add current path
os.chdir(abspath)

from zeroFactory import appFactory

# define self module first
currentModule = sys.modules[__name__]

# define routes
routes = {
		'create': 'createFunc',
		}

address = 'tcp://*:1234'

socketType = sys.argv[1]
if socketType not in ['pull', 'rep']:
	raise TypeError('wrong socket type')

def createFunc(params):
	return 'createdFunction!'

def main():
    app = appFactory.App(routes=routes, currentModule=currentModule,
                         bind=address, socketType=socketType, verbose=True)
    app.run()


if __name__ == "__main__":
    main()

python/test/requester.py
import zmq
import msgpack
import json
import pickle

REQUEST = {
			'jsonrpc': '2.0',
			'method': None,
			'params': None,
			'id': None

}

def setUp(socketType='req'):
	context = zmq.Context()
	socket = context.socket(zmq.REQ)
	socket.connect('tcp://*:1234')
	return socket

def send(socket, d, serializer=msgpack):
	try:
		messagePacked = serializer.dumps(d)
	except:
		messagePacked = serializer.packb(d)

	socket.send(messagePacked)
	msg = socket.recv()

	message = serializer.loads(msg)
	return message

socket = setUp()
OK = 0 

d = {1:2}  # giberrish message to be sent
message = send(socket, d)
expectedResponse = {
					'jsonrpc': '2.0', 
					'id': -1, 
					'error': {
								'message': '"jsonrpc" is required in the message',
								'code': -32600,
								'data': {
											'message': {1: 2}
										}
							}
				}

if message == expectedResponse:
	OK += 1

# Legit Request
d = REQUEST
d['method'] = 'create'
d['id'] = 1

message = send(socket, d)
expectedResponse = {
					'jsonrpc': '2.0',
					'result': 'createdFunction!',
					'id': 1
					}

if message == expectedResponse:
	OK += 1
else:
	print '\n\n %s expected. Got %s instead \n\n' % (expectedResponse, message)

# Request with unknown method
d = REQUEST
d['method'] = 'NotFoundInRoutes'
d['id'] = 2

message = send(socket, d)
expectedResponse = {
					'jsonrpc': '2.0', 
					'id': 2, 
					'error': {
								'message': "'NotFoundInRoutes' is not a valid method",
								'code': -32601,
								'data': {
											'message': {
														'params': None, 
														'jsonrpc': '2.0',
														'method': 'NotFoundInRoutes',
														'id': 2
														}
										}
							}
					}
if message == expectedResponse:
	OK += 1
else:
	print '\n\n %s expected. Got %s instead \n\n' % (expectedResponse, message)


# Request without parameters
d = {'jsonrpc': '2.0'}
d['method'] = 'create'
d['id'] = 3

message = send(socket, d)
expectedResponse = {
					'jsonrpc': '2.0',
					'id': 3, 
					'error': {
								'message': 'params is required in the message',
								'code': -32600,
								'data': {
										'message': {
													'jsonrpc': '2.0',
													'method': 'create',
													'id': 3
													}
										}
							}
					}
if message == expectedResponse:
	OK += 1
else:
	print '\n\n %s expected. Got %s instead \n\n' % (expectedResponse, message)

print OK

python/zeroFactory/appWrapper.py
import zmq

'''
This is a standardized client for zeroFactory apps.
You may or may not want to include this in your apps.
The nice thing about this client wrapper is that it handles disconnects well.

It will basically close and reopen the socket if something bad happens.
'''

JSONRPCMESSAGE = {
                    "jsonrpc": "2.0",
                    "method": None,
                    "params": None,
                    "id": None
                }


class AppWrapper:
    def __init__(self, context, socketType, addresses):
        '''
        Initialize the app you want to connect to by passing in all the details
        '''
        self.context = context
        self.socketType = socketType
        self.addresses = addresses
        self.socket = self.context.socket(socketType)
        self.connect()

    def connect(self):
        '''
        connect to app (zeroFactory or not)
        '''
        if type(self.addresses) == list or type(self.addresses) == tuple:
            for address in self.addresses:
                self.socket.connect(address)
        else:
            self.socket.connect(self.addresses)

    def reopen(self):
        self.socket.close()
        # reinitialize the socket
        self.socket = self.context.socket(self.socketType)
        self.connect()

    def send(self, message):
        '''
        zmq.NOBLOCK is for cases where the receipient does not exist
        '''
        self.socket.send(message, zmq.NOBLOCK)

    def recv(self):
        return self.socket.recv(zmq.NOBLOCK)

    def sendRecv(self, message, timeout):
        '''
        Usage:
        msg = app.sendRecv(messageSent, 1)

        This will automatically block until you have received a reply,
        or until timeout (in seconds)

        For high performance system and patterns (shotgun patterns for example)
        use a low timeout.
        '''
        self.send(message)
        poller = zmq.Poller()
        poller.register(self.socket, zmq.POLLIN)

        try:
            timeout = timeout * 1000
        except TypeError:
            timeout = None
        while True:
            socks = dict(poller.poll(timeout))
            if socks:
                if socks.get(self.socket) == zmq.POLLIN:
                    return self.recv()
            else:
                self.reopen()
                return None


python/zeroFactory/jsonRPCWrapper.py
import msgpack
import json

from zeroFactoryExceptions import *

SUCCESSMESSAGE = {
    "jsonrpc": "2.0",
    "result": None,
    "id": None
}

ERRORMESSAGE = {
    "jsonrpc": "2.0",
    "error": {
        "code": None,
        "message": None,
        "data": None
            },
    "id": None
}


def validateJSONRPCMessage(message):
    '''
    type(message) == dict
    '''
    if type(message) != dict:
        # if it is improper JSON, ValueError will be raised
        try:
            message = msgpack.unpackb(message)
        except:
            try:
                message = json.loads(message)
            except:
                # raise error
                errorString = '''The type of message passed in needs 
                                to be dict, msgpack or json'''
                raise UnknownMessageType(errorString)

    # check if "jsonrpc" == 2.0
    try:
        if message['jsonrpc'] != '2.0':
            errorString = '''JSON RPC version supplied is %s.
                             Only "2.0" is accepted.''' % message['jsonrpc']
            raise UnsupportedJSONRPCVersion(errorString)
    except KeyError:
        errorString = '"jsonrpc" is required in the message'
        raise RequiredKeysMissing(errorString)

    # check if key "method" exists
    # check if key "params" exists
    # check if key "id" exists
    for definedKeys in ['method', 'params', 'id']:
        if definedKeys not in message.keys():
            errorString = '%s is required in the message' % definedKeys
            raise RequiredKeysMissing(errorString)

    return message

python/zeroFactory/appFactory.py
import sys
import os
import zmq.green as zmq
import json
import pickle
import msgpack
import traceback

# fix import and weird WSGI path issues
abspath = os.path.abspath(os.path.dirname(sys.argv[0]))
parentdir = os.path.dirname(abspath)
sys.path.insert(1, parentdir)  # add current path
os.chdir(abspath)

from jsonRPCWrapper import *
from zeroFactoryExceptions import *


class App():

    __slots__ = ['routes', 'currentModule', 'socketType', 'bind', 'format']

    def __init__(self, routes=None, currentModule=None,
                socketType='rep', bind=None, format='msgpack', verbose=None):
        '''
            routes should be a dict. An example:
                routes = {
                        'create-recipe': 'createRecipe',
                        'retrieve-recipe': 'retrieveRecipe',
                        'update-recipe': 'updateRecipe',
                        'delete-recipe': 'deleteRecipe'
                        }

            currentModule MUST be passed in.
            currentModule is defined by the main module calling this class:
                currentModule = sys.modules[__name__]

        '''
        self.routes = routes  # routes to call
        self.currentModule = currentModule  # current module is declared
        self.socketType = socketType  # this is needed for run()
        self.format = format
        self.verbose = verbose
        self.bind = bind

        # format of the messages. Defaults to msgpack
        if self.format == 'json':
            self.serializer = json
        elif self.format == 'pickle':
            self.serializer = pickle
        elif self.format == 'msgpack':
            self.serializer = msgpack

        # check if routes and currentModule exists
        if not routes or not currentModule:
            errorString = '''
                            Routes AND Current Module are REQUIRED.
                            '''
            raise Exception(errorString)

        self._bind()

    def _bind(self):
        # ZeroMQ
        self.context = zmq.Context()  # zmq Context
        if self.socketType == 'rep':
            self.socket = self.context.socket(zmq.REP)
        elif self.socketType == 'pull':
            self.socket = self.context.socket(zmq.PULL)

        try:
            self.socket.bind(self.bind)  # bind address from config
        except Exception as e:
            if e.strerror == 'Invalid argument':
                errorString = '''Bind address format: transport://adress:port'''
            elif e.strerror == 'Address in use':
                errorString = '''Binding address in use. Free up address first'''
            else:
                errorString = '%s' % e
            raise Exception(errorString)
        except TypeError:
            errorString = 'Bind address is required'
            raise Exception(errorString)

    def run(self):
        '''
            message received is in this format (we're using jsonrpc v2):
            see http://www.jsonrpc.org/specification
            message = {
                    "jsonrpc": "2.0",
                    "method": ___,
                    "params": {
                        'data1': ___,
                        'data2': ___,
                        },
                    "id": ___
        '''
        try:
            while True:
                rawMessage = self.socket.recv()

                messageUnpacked = self.serializer.loads(rawMessage)
                if self.verbose:
                    print 'received %s' % messageUnpacked

                # ok, so you unpacked the message... now validate it
                try:
                    message = validateJSONRPCMessage(messageUnpacked)
                except RequiredKeysMissing as e:  # damn, you passed in bad request
                    errorData = {'message': messageUnpacked}
                    errorMessage = '%s' % e
                    messageID = self._getMessageID(messageUnpacked)
                    errorDict = self.error(-32600, errorMessage, errorData, messageID)

                    self.reply(errorDict)
                    # TO DO: LOGGING!
                    # Temporary logging: PRINT!
                    print '%s' % errorDict
                    continue

                except UnknownMessageType as e:
                    errorMessage = '%s' % e
                    errorDict = self.error(-32700, errorMessage)
                    self.reply(errorDict)

                    # TO DO: LOGGING!
                    # Temporary logging: PRINT!
                    print '%s' % errorDict
                    continue

                # congrats! the message validated without errors
                try:
                    method = getattr(self.currentModule, self.routes[message['method']])
                except KeyError as e:  # method not found in the route
                    errorData = {'message': messageUnpacked}
                    errorMessage = '%s is not a valid method' % e
                    messageID = self._getMessageID(messageUnpacked)
                    errorDict = self.error(-32601, errorMessage, errorData, messageID)

                    self.reply(errorDict)

                    # TO DO: LOGGING!
                    # Temporary logging: PRINT!
                    print '%s' % errorDict
                    continue

                # great! you found the method. Now call it!
                result = None  # default for result
                try:
                    result = method(message['params'])
                except KeyError as e:  # this means params are not in message
                    errorData = {'message': messageUnpacked}
                    errorMessage = '%s' % e
                    messageID = self._getMessageID(messageUnpacked)
                    errorDict = self.error(-32602, errorMessage, errorData, messageID)
                    print ('KeyError', e)
                    self.reply(errorDict)

                    # TO DO: LOGGING!
                    # Temporary logging: PRINT!
                    print '%s' % errorDict
                except TypeError as e:
                    errorData = {'message': messageUnpacked}
                    errorMessage = '%s' % e
                    messageID = self._getMessageID(messageUnpacked)
                    errorDict = self.error(-32602, errorMessage, errorData, messageID)
                    print ('TypeError', e)
                    self.reply(errorDict)
                except:
                    print traceback.format_exc()
                    e = sys.exc_info()
                    errorType = e[0]
                    try:
                        errorMessage = e[1][0]
                    except IndexError:
                        errorMessage = e[1]

                    errorData = {'message': messageUnpacked}
                    errorMessage = '%s : %s' % (errorType, errorMessage)
                    messageID = self._getMessageID(messageUnpacked)
                    errorDict = self.error(-32603, errorMessage, errorData, messageID)
                    self.reply(errorDict)

                if self.socketType in ['rep']:
                    if result:  # result should be a dict or None
                        wrappedMessage = SUCCESSMESSAGE
                        wrappedMessage['result'] = result
                        wrappedMessage['id'] = messageUnpacked['id']
                        self.reply(wrappedMessage)
                        if self.verbose:
                            print 'sent %s' % wrappedMessage
                    else:
                        methodCalled = messageUnpacked['method']
                        errorMessage = 'No results found :('
                        errorData = {'message': 'No Results found for method %s' % methodCalled}
                        messageID = self._getMessageID(messageUnpacked)
                        errorDict = self.error(-32404, errorMessage, errorData, messageID)
                        self.reply(errorDict)
        except Exception as e:
            # this shouldn't happen. You're fucked.
            if self.verbose:
                print(traceback.format_exc())
                print("RESTARTING THE APP GRACEFULLY")
            self.close()
            self._bind()
            self.run()

        finally:
            self.close()

    def reply(self, d):
        '''
            This function takes d, which is a RESPONSE dictionary (either SUCCESS or ERROR),
            Transforms it nicely into msgpack or whatnots that has been defined in format
            And then sends it. If it needs to be sent
        '''
        # TO DO: verify that d is a dict first
        messagePacked = self.serializer.dumps(d)
        if self.socketType in ['rep']:  # only reply if it's a reply type socket
            try:
                self.socket.send(messagePacked)
            except:
                e = sys.exc_info()
                errorType = e[0]
                try:
                    errorMessage = e[1][0]
                except IndexError:
                    errorMessage = e[1]
                print('%s : %s' % (errorType, errorMessage))
                pass
        else:
            print(d)

    def error(self, code, errorMessage, data=None, messageID=-1):
        '''
            This function wraps errors. No shit sherlock.
        '''
        errorDict = ERRORMESSAGE
        errorDict['error']['code'] = code  # Invalid Request
        errorDict['error']['message'] = errorMessage
        errorDict['error']['data'] = data
        errorDict['id'] = messageID

        return errorDict

    def _getMessageID(self, message):
        try:
            messageID = message['id']
        except:  # all exceptions! Pokemon! Gotta Catch 'em all
            messageID = -1

        return messageID

    def close(self):
        '''
            To undo any bindings if necessary
        '''
        self.socket.close()
        return 'Socket closed'
